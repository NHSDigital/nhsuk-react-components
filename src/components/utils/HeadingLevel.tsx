import React, { FC, HTMLProps } from 'react';

interface HeadingLevelProps extends HTMLProps<HTMLHeadingElement> {
  headingLevel?: HeadingLevelType;
}

export type HeadingLevelType =
  | 'h1'
  | 'h2'
  | 'h3'
  | 'h4'
  | 'h5'
  | 'h6'
  | 'H1'
  | 'H2'
  | 'H3'
  | 'H4'
  | 'H5'
  | 'H6';

const HeadingLevel: FC<HeadingLevelProps> = ({ headingLevel = 'h3', ...rest }) => {
  switch (headingLevel.toLowerCase()) {
    case 'h1':
      return <h1 {...rest} />;
    case 'h2':
      return <h2 {...rest} />;
    case 'h3':
      return <h3 {...rest} />;
    case 'h4':
      return <h4 {...rest} />;
    case 'h5':
      return <h5 {...rest} />;
    case 'h6':
      return <h6 {...rest} />;
    default:
      console.error(`HeadingLevel: Invalid headingLevel prop: ${headingLevel}`);
      return <h3 {...rest} />;
  }
};

export default HeadingLevel;
