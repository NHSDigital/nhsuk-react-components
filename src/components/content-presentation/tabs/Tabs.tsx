'use client';
import classNames from 'classnames';
import React, { FC, HTMLAttributes, useEffect } from 'react';
import HeadingLevel, { HeadingLevelType } from '@components/utils/HeadingLevel';
import TabsJs from '@resources/tabs';

type TabsProps = HTMLAttributes<HTMLDivElement>;

type TabTitleProps = { children: React.ReactNode; headingLevel?: HeadingLevelType };

type TabListProps = {
  children: React.ReactNode;
};

type TabListItemProps = {
  id: string;
  children: React.ReactNode;
};

type TabContentsProps = {
  id: string;
  children: React.ReactNode;
};

const TabTitle: FC<TabTitleProps> = ({ children, headingLevel = 'h2' }) => (
  <HeadingLevel className="nhsuk-tabs__title" headingLevel={headingLevel}>
    {children}
  </HeadingLevel>
);

const TabList: FC<TabListProps> = ({ children }) => (
  <ul className="nhsuk-tabs__list">{children}</ul>
);

const TabListItem: FC<TabListItemProps> = ({ id, children }) => (
  <li className="nhsuk-tabs__list-item">
    <a className="nhsuk-tabs__tab" href={`#${id}`}>
      {children}
    </a>
  </li>
);

const TabContents: FC<TabContentsProps> = ({ id, children }) => (
  <div className="nhsuk-tabs__panel" id={id}>
    {children}
  </div>
);

interface Tabs extends FC<TabsProps> {
  Title: FC<TabTitleProps>;
  List: FC<TabListProps>;
  ListItem: FC<TabListItemProps>;
  Contents: FC<TabContentsProps>;
}

const Tabs: Tabs = ({ className, children, ...rest }) => {
  useEffect(() => {
    TabsJs();
  }, []);

  return (
    <div className={classNames('nhsuk-tabs', className)} data-module="nhsuk-tabs" {...rest}>
      {children}
    </div>
  );
};

Tabs.Title = TabTitle;
Tabs.List = TabList;
Tabs.ListItem = TabListItem;
Tabs.Contents = TabContents;

export default Tabs;
