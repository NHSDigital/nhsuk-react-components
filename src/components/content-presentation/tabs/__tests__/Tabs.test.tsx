import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Tabs from '../Tabs';
import { HeadingLevelType } from '@util/HeadingLevel';

describe('The tabs component', () => {
  it('Matches the snapshot', () => {
    const { container } = render(
      <Tabs>
        <Tabs.Title>Contents</Tabs.Title>
        <Tabs.List>
          <Tabs.ListItem id="past-day">Past day</Tabs.ListItem>
          <Tabs.ListItem id="past-week">Past week</Tabs.ListItem>
          <Tabs.ListItem id="past-month">Past month</Tabs.ListItem>
        </Tabs.List>

        <Tabs.Contents id="past-day">
          <div>Past day contents go here</div>
        </Tabs.Contents>

        <Tabs.Contents id="past-week">
          <div>Past week contents go here</div>
        </Tabs.Contents>

        <Tabs.Contents id="past-month">
          <div>Past month contents go here</div>
        </Tabs.Contents>
      </Tabs>,
    );

    expect(container).toMatchSnapshot();
  });

  it('Switches the visibility of tabs when clicked', () => {
    const { container } = render(
      <Tabs>
        <Tabs.Title>Contents</Tabs.Title>
        <Tabs.List>
          <Tabs.ListItem id="tab-one">Tab One</Tabs.ListItem>
          <Tabs.ListItem id="tab-two">Tab Two</Tabs.ListItem>
        </Tabs.List>

        <Tabs.Contents id="tab-one">
          <div>Tab one contents go here</div>
        </Tabs.Contents>

        <Tabs.Contents id="tab-two">
          <div>Tab two contents go here</div>
        </Tabs.Contents>
      </Tabs>,
    );

    const firstTabLink = container.querySelector('#tab_tab-one');
    const secondTabLink = container.querySelector('#tab_tab-two');

    expect(
      firstTabLink?.parentElement?.classList.contains('nhsuk-tabs__list-item--selected'),
    ).toEqual(true);
    expect(
      secondTabLink?.parentElement?.classList.contains('nhsuk-tabs__list-item--selected'),
    ).toEqual(false);

    fireEvent.click(secondTabLink!);

    expect(
      firstTabLink?.parentElement?.classList.contains('nhsuk-tabs__list-item--selected'),
    ).toEqual(false);
    expect(
      secondTabLink?.parentElement?.classList.contains('nhsuk-tabs__list-item--selected'),
    ).toEqual(true);
  });

  describe('The tabs title', () => {
    it.each`
      headingLevel
      ${undefined}
      ${'H1'}
      ${'H2'}
      ${'H3'}
      ${'H4'}
    `(
      'Renders the chosen heading level $headingLevel if specified',
      ({ headingLevel }: { headingLevel: HeadingLevelType }) => {
        const { container } = render(
          <Tabs.Title headingLevel={headingLevel}>Test title</Tabs.Title>,
        );

        const title = container.querySelector('.nhsuk-tabs__title');

        expect(title?.nodeName).toEqual(headingLevel ?? 'H2');
      },
    );
  });

  describe('The tab list', () => {
    it('Renders the expected children', () => {
      const { container } = render(
        <Tabs.List>
          <div id="list-contents" />
        </Tabs.List>,
      );

      const listElement = container.querySelector('.nhsuk-tabs__list');

      expect(listElement?.querySelector('#list-contents')).toBeTruthy();
    });
  });

  describe('The tab list item', () => {
    it('Sets the href to be the passed in id prop', () => {
      const { container } = render(
        <Tabs.ListItem id="test-id">
          <div id="list-item-contents" />
        </Tabs.ListItem>,
      );

      expect(container.querySelector('.nhsuk-tabs__tab')?.getAttribute('href')).toBe('#test-id');
    });

    it('Renders the expected children', () => {
      const { container } = render(
        <Tabs.ListItem id="test-id">
          <div id="list-item-contents" />
        </Tabs.ListItem>,
      );

      const tabElement = container.querySelector('.nhsuk-tabs__tab');

      expect(tabElement?.querySelector('#list-item-contents')).toBeTruthy();
    });
  });

  describe('The tab contents', () => {
    it('Renders the expected children', () => {
      const { container } = render(
        <Tabs.Contents id="test-contents">
          <div id="tab-contents" />
        </Tabs.Contents>,
      );

      const tabElement = container.querySelector('#test-contents');

      expect(tabElement?.querySelector('#tab-contents')).toBeTruthy();
    });
  });
});
