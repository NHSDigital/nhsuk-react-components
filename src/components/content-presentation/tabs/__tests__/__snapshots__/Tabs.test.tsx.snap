// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The tabs component Matches the snapshot 1`] = `
<div>
  <div
    class="nhsuk-tabs"
    data-module="nhsuk-tabs"
  >
    <h2
      class="nhsuk-tabs__title"
    >
      Contents
    </h2>
    <ul
      class="nhsuk-tabs__list"
      role="tablist"
    >
      <li
        class="nhsuk-tabs__list-item nhsuk-tabs__list-item--selected"
        role="presentation"
      >
        <a
          aria-controls="past-day"
          aria-selected="true"
          class="nhsuk-tabs__tab"
          href="#past-day"
          id="tab_past-day"
          role="tab"
          tabindex="0"
        >
          Past day
        </a>
      </li>
      <li
        class="nhsuk-tabs__list-item"
        role="presentation"
      >
        <a
          aria-controls="past-week"
          aria-selected="false"
          class="nhsuk-tabs__tab"
          href="#past-week"
          id="tab_past-week"
          role="tab"
          tabindex="-1"
        >
          Past week
        </a>
      </li>
      <li
        class="nhsuk-tabs__list-item"
        role="presentation"
      >
        <a
          aria-controls="past-month"
          aria-selected="false"
          class="nhsuk-tabs__tab"
          href="#past-month"
          id="tab_past-month"
          role="tab"
          tabindex="-1"
        >
          Past month
        </a>
      </li>
    </ul>
    <div
      aria-labelledby="tab_past-day"
      class="nhsuk-tabs__panel"
      id="past-day"
      role="tabpanel"
    >
      <div>
        Past day contents go here
      </div>
    </div>
    <div
      aria-labelledby="tab_past-week"
      class="nhsuk-tabs__panel nhsuk-tabs__panel--hidden"
      id="past-week"
      role="tabpanel"
    >
      <div>
        Past week contents go here
      </div>
    </div>
    <div
      aria-labelledby="tab_past-month"
      class="nhsuk-tabs__panel nhsuk-tabs__panel--hidden"
      id="past-month"
      role="tabpanel"
    >
      <div>
        Past month contents go here
      </div>
    </div>
  </div>
</div>
`;
